### PR TITLE
Integrating third-party API Services

### PR DESCRIPTION
**Pull request by wzdnzd**
+ Integrating third-party services and adding new providers
+ Fix the bug of tooltip positioning abnormally when switching settings page
image

### CODE REVIEW

**Comment by yetone** [comment]
> > This is a great job! I have a question, if a user updates to this version will he suddenly not be able to translate because we didn't migrate his old settings to the new one
> 
> That's correct, but users only need to refill the API Key, which is very easy.

But this will hurt the user experience, can we make an automatic migrate logic?

**Comment by wzdnzd** [comment]
> > > This is a great job! I have a question, if a user updates to this version will he suddenly not be able to translate because we didn't migrate his old settings to the new one
> > 
> > 
> > That's correct, but users only need to refill the API Key, which is very easy.
> 
> But this will hurt the user experience, can we make an automatic migrate logic?

After upgrading, users may be affected in their usage, but usually only need to re-enter the API key, which is relatively simple. Adding some one-time code for this purpose may make the code less concise and clear. Of course, it's up to you to decide, as the implementation of the code is relatively simple.

**Comment by yetone** [comment]
Make sense. I need some time to verify the compatibility of this PR in Tauri.

File: src/common/subscribe.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                }
+                modelIndexes.set(item.id, item)
+            }
+
+            model = modelIndexes.get('gpt-4') || modelIndexes.get(defaultAPIModel) || models[0]
+        }
+
+        return { authCode: authCode, model: model }
+    } catch {
+        return undefined
```
Comment (yetone) [inline src/common/subscribe.ts:160]: What error will be caught by this? Ignoring certain error messages may result in difficulties during debugging.

Comment (yetone) [reply src/common/subscribe.ts:160]: Perhaps we can eliminate the try-catch block.

File: src/common/subscribe.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            if (auths && typeof auths === 'object') {
+                Object.entries(auths).map(([k, v]) => {
+                    const authCodeModel: AuthCodeModel = v as AuthCodeModel
+                    if (k && authCodeModel.authCode) {
+                        this.cache.set(k, authCodeModel)
+                    }
+                })
+            }
+        } catch {
+            console.error(`cache auth code and model from settings error`)
```
Comment (yetone) [inline src/common/subscribe.ts:215]: Will `this.cache.set` cause an error? Maybe we can remove the try-catch block.

File: src/common/components/Settings.tsx [hunk 3]
```diff
@@ -982,18 +1033,87 @@ export function InnerSettings({ onSave }: IInnerSettingsProps) {
...
+                setValues(primitiveForm)
+                setPrevValues(primitiveForm)
             })()
         }
     }, [isTauri, settings])
 
-    const onChange = useCallback((_changes: Partial<ISettings>, values_: ISettings) => {
-        setValues(values_)
-    }, [])
+    const validateSettings = useCallback(
```
Comment (yetone) [inline src/common/components/Settings.tsx:1044]: We should use this to do the validation: https://github.com/openai-translator/openai-translator/blob/41ef88b98227e65a178b43acda2a84e3bbb149b1/src/common/components/Form/form.ts#L49

**Comment by OverflowCat** [comment]
https://github.com/openai-translator/openai-translator/assets/20166026/f7da0877-0d94-41e3-b21e-a6e7a54b77bf

Is this the expected behavior?

File: src/common/subscribe.ts (hunk 1 above)
Comment (wzdnzd) [reply src/common/subscribe.ts:160]: There should be no errors theoretically, and I will remove it later.

File: src/common/subscribe.ts (hunk 2 above)
Comment (wzdnzd) [reply src/common/subscribe.ts:215]: There should be no errors theoretically, and I will remove it later.

File: src/common/components/Settings.tsx (hunk 3 above)
Comment (wzdnzd) [reply src/common/components/Settings.tsx:1044]: If I put the validation in each FormItem, how should I write the validator to achieve the current rules (mainly that different FormItems will affect each other)? Do you have any suggestions or examples?

**Comment by wzdnzd** [comment]
> Is this the expected behavior?

Can't switch provider? I haven't tried it on desktop, but I've been using it in Chrome for a month and haven't encountered this situation. Do you have more detailed information?

@yetone Have you tried Tauri? Do you have this issue there?

**Comment by yetone** [comment]
> 

Oh sorry, I am not working on Tauri testing, because I am busy with the react-syntax-highlighter bug these days

**Comment by yetone** [comment]
Yes, I also have this issue in Tauri

https://github.com/openai-translator/openai-translator/assets/1206493/e0e3c2da-b9e1-4682-ad90-675b5b2ab5b3

**Comment by wzdnzd** [comment]
@yetone @OverflowCat I specially set up a Tauri development environment for testing today, and it works fine on my Windows system. As I don't have any Linux or MacOS devices at hand temporarily, could you please help me locate the problem?😂😂😂

https://github.com/openai-translator/openai-translator/assets/8565764/ac2b57f8-5ed0-4270-a4dc-dc8e919470fc

**Pull request closed without merging**